### DISCUSSION TITLE
How to configure Polly v8 with the Ocelot 23.2.2?

### DISCUSSION THREAD

**Discussion by anonimousspider** [ThreeMammals/Ocelot#2083] [Q&A] [answered]
**In Polly v8, we have observed that the circuit breaker functionality is not working as expected. We have experienced issues where the circuit breaker does not trip or reset correctly, leading to inconsistent fault handling. Conversely, in Polly v7, the circuit breaker also exhibits problems, failing to operate smoothly in certain scenarios. 
Circuit Breaker Not Tripping** 

### Example
Suppose you have configured an Ocelot route with a circuit breaker policy to trip after three consecutive HTTP request failures:
Route Configuration:
```json
{
  "Routes": [
    {
      "DownstreamPathTemplate": "/api/StudentIdentity/Verify ",
      "DownstreamScheme": "https",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost", "Port": 5001
        }
      ],
      "UpstreamPathTemplate": " api/student/Verify ",
      "UpstreamHttpMethod": [ "Get" ],
      "QoSOptions": {
        "ExceptionsAllowedBeforeBreaking": 3,
        "DurationOfBreak": 10000,
        "TimeoutValue": 5000
      }
    }
  ]
}
```
Configuration in Program.cs for V7:
```C#
services.AddLogging(logging =>
 {
     logging.ClearProviders();
     logging.AddConsole();
                            
 });
 services.AddJwtAuthentication();
 services.AddOcelot().AddPollyV7();
```
Configuration in program.cs using Polly V8:
```C#
services.AddLogging(logging =>
 {
     logging.ClearProviders();
     logging.AddConsole();
                            
 });
 services.AddJwtAuthentication();
 services.AddOcelot().AddPolly();
```

### Expected Behaviour
During runtime, the downstream service fails five consecutive times. The circuit breaker should trip and stop the request to be forwarded , and move to open state as per configured. 

### Actual behavior
During runtime, the downstream service fails five consecutive times. Despite this, the circuit breaker does not trip and continues to forward requests, instead of moving to the open state after three failures as configured.

### Scenario 1
When service response is delayed from the service configured in ocelot.
**PollyV7:** Its working fine 
**PollyV8:** on applying pollyv8 in the gateway not Working.

### Scenario 2
When service is giving 503 error response .
**Polly v7** : Its working Fine 
**Polly v8:** Not working.

### Scenario 3
When Service is giving 404 error response.
**Polly v7 :** working fine.
**Polly v8:** Not working.

---
Can you please suggest, how to configure Polly v8 with the current Ocelot version 23.2.2?

**Comment by raman-m** [comment]
@anonimousspider, thank you for your question! Could you please provide your full name?
Welcome to the world of Ocelot! 🐯

It seems you're experiencing issues with the V7 provider. However, since the QoS package has been updated to the latest Polly version (V8 syntax), we advise using only Ocelot's `AddPolly` which is V8.
For your information, we plan to completely phase out the V7 provider in Ocelot v23.3 (the current release).
For more details, refer to PR #2079.

**Reply by raman-m** [reply]
@RaynaldM, your client is here! 😉 Could you provide further advice to the developer, please?
Are there similar acceptance tests that are more closely related to the problem?

**Comment by anonimousspider** [comment]
can you please guide me on how to configure polly v8 correctly.

**Reply by raman-m** [reply, marked as answer]
### Official Polly documentation
- [Meet Polly: The .NET resilience library](https://www.pollydocs.org/)
- [Resilience pipelines](https://www.pollydocs.org/pipelines/index.html)

### Official Ocelot documentation
- [Quality of Service](https://ocelot.readthedocs.io/en/latest/features/qualityofservice.html)

What's your full name?
What's your LinkedIn?

**Comment by RaynaldM** [comment]
Hi Anonimous, I'm trying to reproduce your case.
First, according to [this list](https://github.com/ThreeMammals/Ocelot/blob/ee1fb97fcf08c8785a45a5576591f0bb95baeda7/src/Ocelot.Provider.Polly/PollyQoSProviderBase.cs#L8) , 404 is not a circuit-breaker launcher, you could use you own ServerErrorCodes to add it

**Reply by RaynaldM** [reply]
The unit test that checks that the circuit-breaker is working correctly is [here](https://github.com/RaynaldM/Ocelot/blob/42ac3ca51829006bbe2e63bc684ab3a3caaccf1e/test/Ocelot.UnitTests/Polly/PollyQoSProviderTests.cs#L117), I've run it, and it validates that the process is running correctly :
![image](https://github.com/ThreeMammals/Ocelot/assets/606401/6c7762e2-ff47-444b-9f32-1553af441ed4)

_I don't know what more I can do for you_  in fact, no :
I have found a probable issue with strategies order, PR is running #2086 (issue #2085)

**Reply by raman-m** [reply]
OK, PR #2086 will officially close this discussion...

**Comment by anonimousspider** [comment]
Hi @RaynaldM as you mentioned that the circuit breaker is working properly but when i am configuring the Polly v8 its not working properly. For better understanding I am sharing you my project can you please guide whether I have done mistakes in configuring PollyV8

**Reply by raman-m** [reply]
You must upload the solution as a new repo to your GiHub account❗

**Reply by anonimousspider** [reply]
Hi @raman-m as you have guided I have committed the code to the repository below mentioned is the link to the repo.
https://github.com/anonimousspider/ExampleGatewayProj
Can you please guide me with the configuration of Polly V8..!!

**Reply by raman-m** [reply]
> [WeatherForecast](https://github.com/anonimousspider/OcelotProject/blob/main/GatewayProject/WeatherForecast.cs) 🤣 
> [WeatherForecastController](https://github.com/anonimousspider/OcelotProject/blob/main/GatewayProject/Controllers/WeatherForecastController.cs)

Why do you have unnecessary controllers in gateway project? That's absolutely wrong from the view point of Microservices architecture!
All data controllers must be moved to downstream service project!

Also, the [ocelot.json](https://github.com/anonimousspider/OcelotProject/blob/main/GatewayProject/ocelot.json) file is overly extensive and contains superfluous JSON properties. It should be condensed to a shorter version to avoid errors in route configuration. Only the necessary JSON sections and options need to be defined.

I'm asking once again...
- What's your full name?
- What's your LinkedIn ?

I will postpone any feedback until you answer these questions. I'm not fond of conversing with anonymous individuals.

**Comment by RaynaldM** [comment]
For me, your sample seems to be correct. I don't unsderstand where is the issue

**Reply by raman-m** [reply]
@RaynaldM 
This user is blocked! The thread will be closed. This is my decision...
Please disregard all anonymous inquiries. Thank you!

**Comment by anonimousspider** [comment]
Hi @RaynaldM  sorry there was some conflict happend in the previous repo. 
Here I have created the new repo. can you please guide me ??
I have already share you the concerned problem. even logs are also there for the same. Is there any thing to solve the problem.
Repo link : https://github.com/anonimousspider/ExampleGatewayProj
